consensus/src/chained_bft/event_processor.rs

```diff
@@ -60,6 +60,56 @@ use std::{
 };
 use termion::color::*;
 
+pub enum UnverifiedEvent<T> {
+    ProposalMsg(Box<ProposalMsg<T>>),
+    VoteMsg(Box<VoteMsg>),
+    SyncInfo(Box<SyncInfo>),
+}
+
+impl<T: Payload> UnverifiedEvent<T> {
+    pub fn verify(self, validator: &ValidatorVerifier) -> Result<VerifiedEvent<T>> {
+        Ok(match self {
+            UnverifiedEvent::ProposalMsg(p) => {
+                p.verify(validator)?;
+                VerifiedEvent::ProposalMsg(p)
+            }
+            UnverifiedEvent::VoteMsg(v) => {
+                v.verify(validator)?;
+                VerifiedEvent::VoteMsg(v)
+            }
+            UnverifiedEvent::SyncInfo(s) => {
+                s.verify(validator)?;
+                VerifiedEvent::SyncInfo(s)
+            }
+        })
+    }
+
+    pub fn epoch(&self) -> u64 {
```

Comment:
maybe make each of these classes implement a trait on some ConsensusMsg trait that exposes both a verify and epoch method? So that you can then make a generic function here on that trait?